### ISSUE TITLE
JDK lookups should include sdkman installed JDKs

### ISSUE THREAD

**Issue by jeffmaury** [jbangdev/jbang#1476]
**Is your feature request related to a problem? Please describe.**
When a specific version of Java is required for a script, the JDK installed by sdkman should be looked

**Describe the solution you'd like**
I had Java 17.0.3 installed by sdkman and ran https://raw.githubusercontent.com/maxandersen/gh-jtriage/main/jtriage.java but jbang says Downloading JDK 16.
Not mentioning permission issues on Windows

**Describe alternatives you've considered**

**Additional context**

**Comment by quintesse** [comment]
@jeffmaury can you tell us exactly which JDK you have installed using sdkman?

And if possible can you see if that installed JDK has a `release` file in its root folder? And if so could you paste its contents?

**Comment by maxandersen** [comment]
@jeffmaury is your case you don't have that java 17 from sdkman in path and you expect jbang to go search on local disk for alternative sdk installs?

If yes then I'm not convinced that will provide a consistent behavior as sdkman has no notion of ordering of its jdks. 

You can if you want tell jbang about specific version of jdk to use:

"jbang jdk install 17 `sdk home java 17.0.4.1-tem`"

**Comment by jeffmaury** [comment]
> @jeffmaury can you tell us exactly which JDK you have installed using sdkman?
> 
> And if possible can you see if that installed JDK has a `release` file in its root folder? And if so could you paste its contents?
> 
> Edit: Oh, and to be sure, if you type `java -version` in a terminal you get that 17..0.3 version, right?

I was running JBang with Java11.
I had:

- 8.0.302-tem
- 11.0.16.1-tem
- 17.0.3-tem
- 21.3.0.r11-grl

installed with sdkman.
17.0.3-tem has the following `release` file:
```

IMPLEMENTOR="Eclipse Adoptium"
IMPLEMENTOR_VERSION="Temurin-17.0.3+7"
JAVA_VERSION="17.0.3"
JAVA_VERSION_DATE="2022-04-19"
LIBC="default"
MODULES="java.base java.compiler java.datatransfer java.xml java.prefs java.desktop java.instrument java.logging java.management java.security.sasl java.naming java.rmi java.management.rmi java.net.http java.scripting java.security.jgss java.transaction.xa java.sql java.sql.rowset java.xml.crypto java.se java.smartcardio jdk.accessibility jdk.internal.jvmstat jdk.attach jdk.charsets jdk.compiler jdk.crypto.ec jdk.crypto.cryptoki jdk.crypto.mscapi jdk.dynalink jdk.internal.ed jdk.editpad jdk.hotspot.agent jdk.httpserver jdk.incubator.foreign jdk.incubator.vector jdk.internal.le jdk.internal.opt jdk.internal.vm.ci jdk.internal.vm.compiler jdk.internal.vm.compiler.management jdk.jartool jdk.javadoc jdk.jcmd jdk.management jdk.management.agent jdk.jconsole jdk.jdeps jdk.jdwp.agent jdk.jdi jdk.jfr jdk.jlink jdk.jpackage jdk.jshell jdk.jsobject jdk.jstatd jdk.localedata jdk.management.jfr jdk.naming.dns jdk.naming.rmi jdk.net jdk.nio.mapmode jdk.random jdk.sctp jdk.security.auth jdk.security.jgss jdk.unsupported jdk.unsupported.desktop jdk.xml.dom jdk.zipfs"
OS_ARCH="x86_64"
OS_NAME="Windows"
SOURCE=".:git:9190d7016e2b"
BUILD_SOURCE="git:955033f"
BUILD_SOURCE_REPO="https://github.com/adoptium/temurin-build.git"
SOURCE_REPO="https://github.com/adoptium/jdk17u.git"
FULL_VERSION="17.0.3+7"
SEMANTIC_VERSION="17.0.3+7"
BUILD_INFO="OS: Windows Server 2012 R2 Version: 6.3"
JVM_VARIANT="Hotspot"
JVM_VERSION="17.0.3+7"
IMAGE_TYPE="JDK"
```

**Comment by maxandersen** [comment]
Yes. So by you having java 11 in path that is what jbang will use and if does not fit it will fetch one of the adoptium installs via foojay api. 

there is no (sane) way for us to pick a reliable java version from sdkman as it has no ordering nor priority and multiple product variations.

**Comment by jeffmaury** [comment]
Why don't you search in `$HOME/.sdkman/candidates/java` based on the `release` file @quintesse mentioned ?

**Comment by jeffmaury** [comment]
> @jeffmaury here is what is in my sdkman:
> 
> ```
> 8.0.262.hs-adpt
> 8.0.265.hs-adpt
> 8.0.322-zulu
> 11.0.2-open
> 11.0.9.hs-adpt
> 11.0.11.hs-adpt
> 14.0.2-open
> 14.0.2-sapmchn
> 14.0.2-zulu
> 14.0.2.fx-librca
> 14.0.2.hs-adpt
> 14.0.2.j9-adpt
> 16.0.2-sem
> 16.ea.31-open
> 16.ea.36-open
> 17-open
> 17.0.4.1-tem
> 18-open
> 19-open
> 19.ea.1.lm-open
> 19.ea.12-open
> 20.1.0.r11-grl
> 21.0.0.r11-grl
> 21.1.0.r16-grl
> 21.2.0.r16-grl
> 22.1.0.r17-grl
> current -> 17.0.4.1-tem
> ```
> 
> which version do you suggest we should we pick for Java 16+ ?

I would check current first (but it's likely going to fail as it's probably the one used to run JBang) then retrieve the versions from other candidates, sort them according to semver and select the first compatibile

**Comment by maxandersen** [comment]
@jeffmaury that is slow, and how to choose between using graal vs open vs temurin vs librca ... it just not worth applying this complexity on *all* jbang users.

you can write a script that syncs sdkman installs into jbang if users want that. Happy to add that to jbang-catalog if you wanna contribute it?

**Comment by jeffmaury** [comment]
> @jeffmaury that is slow, and how to choose between using graal vs open vs temurin vs librca ... it just not worth applying this complexity on _all_ jbang users.
That is much faster than downloading a JDK when you have one on your machine 😄 
> 
> you can write a script that syncs sdkman installs into jbang if users want that. Happy to add that to jbang-catalog if you wanna contribute it?
Will do

**Comment by quintesse** [comment]
This was implemented by https://github.com/jbangdev/jbang/pull/1500

But for now it still only works when explicitly enabling the sdkman provider. I that perhaps something we should think about changing @maxandersen and make scoop and sdkman enabled by default?

**Comment by quintesse** [comment]
> which version do you suggest we should we pick for Java 16+ ?

That's a good point, but basically we have the same issue when no 16+ JDK is currently installed and we simply select one for the user. In the case of 16+ we'd currently install the most recent temurin version 17 that we can obtain from the Disco API.

That's a completely arbitrary decision. So couldn't we make a similar arbitrary decision in the case of a long list of locally installed JDKs?

We could perhaps use Maven's version comparator to select the latest version? (Preferring non-EA over EA)
Although any algorithm would do, even simply alphabetical. In the end the only thing we care about is that it always chooses the same one as long as the list of installed JDKs doesn't change.

But to be honest we need to handle this issue in some way if we ever want to implement [Make it possible to have multiple JDKs installed of the same version](https://github.com/jbangdev/jbang-devkitman/issues/46).

**Comment by maxandersen** [comment]
Most recent temurin is inifnitly more deterministic than doing a version string comparison across possible multiple versions. 

I wouldn't compare it to handling multiple jdks as here you would have some additional data to guide the selection wouldn't you ?

**Comment by quintesse** [comment]
> Most recent temurin is inifnitly more deterministic than doing a version string comparison across possible multiple versions.

Right now nothing is deterministic about which version your code is run with if you say "//JAVA 16+". What version gets installed today might well be different from what gets installed tomorrow (it might even go from EA to GA, who knows). 

But even more important, you say "what if I have this long list of sdkman jdks installed?", well, what if you have a long list of jbang jdks installed? It will still run the most recent version. Which version is that? Well that depends on the jdks you got installed!

And in the same way that you don't know which exact version you'll be getting from Temurin it doesn't matter if you do string comparisons across versions, the only important thing is that it always chooses the same version! (And not version 17.1.1 now and 17.2.3 in the next run)

Of course if you install a new Jdk in between runs things might change, but the same happens now! If you use "//JAVA 16+" and have  16 and 21 installed it will use 21, but if you then install 24 it will switch to that one.

> 
> I wouldn't compare it to handling multiple jdks as here you would have some additional data to guide the selection wouldn't you ?

What additional data would that be?

What if I simply want to install 2 slightly different versions from Temurin because there's a bugfix in the Jdk I want to test? Or I want to compare Temurin to the same version of Zulu? I want JBang to be able to install them, and when I'm doing my testing I can specifically select the one I want to use (either by using `jbang jdk default` or by selecting it's specific unique id, etc) but I still want `//JAVA 16+` to work as well and select one of them. Which one? Well, is that important? They should both work. And again, as long as JBang always selects the same one I won't get any weird stuff (although theoretically even that shouldn't cause any problems because all Jdks have to pass the Tck)

Edit: and then there's still [Allow to set a default JDK per version](https://github.com/jbangdev/jbang-devkitman/issues/47) which is a proposal to allow you to choose the default JDK to use _per version_.

**Comment by quintesse** [comment]
To short-circuit this discussion a bit, because we can go on and on about all the little details, I have just one simple question:

_what is the problem with not knowing which exact JDK will run your code? what problems do you foresee that will confuse users or even make things fail?_
